DICOMPlugins/DICOMM3DPlugin.py

```diff
@@ -0,0 +1,641 @@
+from __future__ import absolute_import
+import glob
+import os
+import json
+import vtk
+import io
+import shutil
+import vtkSegmentationCorePython as vtkSegmentationCore
+import logging
+import pydicom
+
+from base.DICOMPluginBase import DICOMPluginBase
+
+import slicer
+from DICOMLib import DICOMLoadable
+
+from SlicerDevelopmentToolboxUtils.mixins import ModuleLogicMixin
+from SlicerDevelopmentToolboxUtils.constants import DICOMTAGS
+from six.moves import range
+
+
+#
+# This is the plugin to handle translation of DICOM M3D objects
+# M3D stands for Model for 3D Manufacturing.
+# DICOM M3D objects can be for example .STL files encoded into a DICOM M3D object.
+#
+class DICOMM3DPluginClass(DICOMPluginBase):
+
+  def __init__(self):
+    super(DICOMM3DPluginClass,self).__init__()
+    self.loadType = "DICOMM3D"
+
+  def examineFiles(self,files):
+    """ Returns a list of DICOMLoadable instances
+    corresponding to ways of interpreting the
+    files parameter.
+    """
+    loadables = []
+
+    for candidateFile in files:
+
+      uid = slicer.dicomDatabase.fileValue(candidateFile, self.tags['instanceUID'])
+      if uid == '':
+        return []
+
+      desc = slicer.dicomDatabase.fileValue(candidateFile, self.tags['seriesDescription'])
+      if desc == '':
+        desc = "Unknown"
+
+      #read EncapsulatedDocument for candidateFile
+      docFile = self.getEncapsulatedDocument(candidateFile)
+
+      #read EncapsulatedDocumentLength for candidateFile
+      docLengthFile = self.getEncapsulatedDocumentLength(candidateFile)
+
```

Comment:
These methods both read the full file, which may take lots of time.  I suggest you refactor and maybe combine these into a single method that does both operations.

But also, do you really need to read the full file just to know if it's an M3D file?  The load step should be as lightweight as possible so that you only invest time reading the data if you are sure it's the right interpretation for the files (e.g. it has the highest confidence or the user has specifically selected it).

Reply:
> These methods both read the full file, which may take lots of time. I suggest you refactor and maybe combine these into a single method that does both operations.

I combined both of these methods, also they are not necessary to determine if it's an M3D object, only the modality tag is the indicator for M3D.

Instead of creating loadable attributes for EncapsulatedDocument and EncapsulatedDocumentLength inside examineFiles, I call these methods directly into the load function.